File: code/ts/test/plugin_host/plugin_host.stage0.red.test.ts
import { describe, expect, it } from 'vitest';
import request from 'supertest';
import {
  ModularApi,
  type Plugin,
  type PluginHost,
  type PluginManifest,
} from '../../src';

describe('Stage 0 red baseline - plugin host', () => {
  it('exposes plugin registration and public plugin types', () => {
    const api = new ModularApi({ basePath: '/api' });
    const plugin = new ProbePlugin();

    expect(api.plugin(plugin)).toBe(api);
    expect(plugin.manifest.id).toBe('acme.echo');
  });

  it('mounts plugin routes only under the shared basePath', async () => {
    const api = new ModularApi({ basePath: '/api' }).plugin(new ProbePlugin());
    const server = await api.serve({ port: 0 });

    try {
      await request(server).get('/api/plugin-probe').expect(200);
      await request(server).get('/plugin-probe').expect(404);
    } finally {
      await new Promise<void>((resolve, reject) => {
        server.close((error) => {
          if (error) {
            reject(error);
            return;
          }

          resolve();
        });
      });
    }
  });
});

class ProbePlugin implements Plugin {
  readonly manifest: PluginManifest = {
    id: 'acme.echo',
    displayName: 'Echo Probe',
    version: '0.1.0',
    hostApiVersion: '>=0.1.0 <0.2.0',
  };

  setup(host: PluginHost): void {
    host.registerRoute({
      id: 'probe-route',
      method: 'GET',
      path: '/plugin-probe',
      visibility: 'custom',
      handler: () => ({ status: 200, body: { ok: true } }),
    });
  }
}

File: code/dart/modular_api/test/plugin_host/plugin_host_stage0_red_test.dart
import 'dart:io';

import 'package:modular_api/modular_api.dart';
import 'package:test/test.dart';

void main() {
  group('Stage 0 red baseline - plugin host', () {
    test('ModularApi should expose plugin registration and public plugin types', () {
      final api = ModularApi(basePath: '/api');

      final ProbePlugin plugin = ProbePlugin();

      expect(api.plugin(plugin), same(api));
      expect(plugin.manifest.id, 'acme.echo');
    });

    test('plugin routes should resolve under the shared basePath', () async {
      final api = ModularApi(basePath: '/api')..plugin(ProbePlugin());

      final server = await api.serve(port: 0);
      addTearDown(server.close);

      final client = HttpClient();
      final prefixed = await client.getUrl(Uri.parse('http://127.0.0.1:${server.port}/api/plugin-probe'));
      final prefixedResponse = await prefixed.close();
      expect(prefixedResponse.statusCode, 200);

      final root = await client.getUrl(Uri.parse('http://127.0.0.1:${server.port}/plugin-probe'));
      final rootResponse = await root.close();
      expect(rootResponse.statusCode, 404);
    });
  });
}

class ProbePlugin implements Plugin {
  @override
  final PluginManifest manifest = PluginManifest(
    id: 'acme.echo',
    displayName: 'Echo Probe',
    version: '0.1.0',
    hostApiVersion: '>=0.1.0 <0.2.0',
  );

  @override
  void setup(PluginHost host) {
    host.registerRoute(
      PluginRoute(
        id: 'probe-route',
        method: 'GET',
        path: '/plugin-probe',
        visibility: 'custom',
        handler: (_) => Response.ok('ok'),
      ),
    );
  }
}